### PR TITLE
modules: hal_silabs: Add Silabs Gecko HAL for EFR32BG22C224F512IM40 SoC

### PR DESCRIPTION
**Pull request by lpawelcz**
Hi, this PR supersedes https://github.com/zephyrproject-rtos/hal_silabs/pull/13. It is a minimal code that should be added to hal_silabs module in order to get initial support for `EFR32BG22` SoCs. Added code is free of MSLA license which was a problem in https://github.com/zephyrproject-rtos/hal_silabs/pull/13.
Although it is a minimal version of the previous PR, we still need to import code and headers for sleeptimer, power_manager, hfxo_manager together with device_init because we do need all of that for our initial support PR in Zephyr.

Origin: Silicon Labs Gecko SDK
URL:
https://www.silabs.com/products/development-tools/software/simplicity-studio Version: 3.1.2
Purpose: board configuration files are added.
License: Zlib
Maintained-by: External

Co-authored-by: Pawel Czarnecki <pczarnecki@antmicro.com>
Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>
Signed-off-by: Paweł Czarnecki <pczarnecki@antmicro.com>

### CODE REVIEW

**Comment by fkokosinski** [comment]
Hi @chrta @mnkp @galak, sorry for pinging you again, but could you take a look at these changes?

**Review by fkokosinski — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]
LGTM

**Comment by sateeshkotapati** [comment]
Please take all the files from GSDK, for e.g., emlib source code and include files) to maintain consistency.

**Review by galak — CHANGES_REQUESTED** [review]
can you remove the board code.  Any board code should be handled in Zephyr.

**Comment by lpawelcz** [comment]
> can you remove the board code. Any board code should be handled in Zephyr.

@galak I removed the whole `board` directory. In order to do that I had to create additional [header](https://github.com/zephyrproject-rtos/zephyr/blob/723b591023be47cfecb1e39a2da6b6e225fc56a7/boards/arm/efr32bg_sltb010a/sl_device_init_hfxo_config.h) for defines used in HAL for hfxo initialization in [Zephyr PR](https://github.com/zephyrproject-rtos/zephyr/pull/50352) under board directory. This header is [included](https://github.com/zephyrproject-rtos/hal_silabs/pull/19/files#diff-e1d62ea151c2cec626118ce0399cf6d9e08064d4b6aca929e2dd3f95380a1641R44) in HAL `CMakeLists.txt` through `ZEPHYR_BASE` environment variable. We can see similar solution in [hal_atmel ](https://github.com/zephyrproject-rtos/hal_atmel/blob/master/asf/common/components/wifi/winc1500/CMakeLists.txt#L3).

Please let me know if that change is OK.

File: gecko/CMakeLists.txt [hunk 1]
```diff
@@ -10,11 +10,38 @@
...
+  set(SILABS_GECKO_BOARD ${CONFIG_BOARD})
+else()
+  string(SUBSTRING ${CONFIG_BOARD} ${SILABS_GECKO_BRD_INDEX} -1 SILABS_GECKO_BOARD)
+endif()
+
 set(SILABS_GECKO_PART_NUMBER ${CONFIG_SOC_PART_NUMBER})
 
+if (${CONFIG_BOARD} STREQUAL "efr32bg_sltb010a")
+  set(SILABS_GECKO_BOARD "brd4184b")
+endif()
```
Comment (kmeinhar) [inline gecko/CMakeLists.txt:26]: To me it looks like setting `SILABS_GECKO_BOARD` only effects which `sl_device_init_clocks.c` is compiled in.
To me the content in `brd4184b/sl_device_init_clocks.c` seems to be specific to SoC and not board specific.

We defined a new board which means `${CONFIG_BOARD}` will not be `efr32bg_stlb010a`. That means without altering this `CMakeLists.txt` our project can not compile.

Comment (lpawelcz) [reply gecko/CMakeLists.txt]: Right, I removed the board specific code (`sl_device_init_clocks.c`) from HAL and implemented a separate clock init function in Zephyr PR as part of the board support. Now `SILABS_GECKO_BOARD` is not used anywhere so I deleted it. I believe that, from zephyr perspective, things like clock configuration are specific to given platform, not to whole SoC. That is why I placed the code in board init file.

File: gecko/CMakeLists.txt [hunk 2]
```diff
@@ -10,11 +10,28 @@
...
+  service/power_manager/config
+  service/hfxo_manager/src
+  service/hfxo_manager/inc
+  service/hfxo_manager/config
+  common/inc
+  service/device_init/inc
+  service/device_init/config/s2
+  service/device_init/config/s2/sdid205
+  ${ZEPHYR_BASE}/boards/arm/${CONFIG_BOARD}
   )
```
Comment (kmeinhar) [inline gecko/CMakeLists.txt:35]: I think `BOARD_ROOT` is a better fit here. But this variable is a list.
A potential solution could be:
```suggestion
  )
foreach(root ${BOARD_ROOT})
  if(IS_DIRECTORY "${root}/boards/arm/${CONFIG_BOARD}")
    zephyr_include_directories(${root}/boards/arm/${CONFIG_BOARD})
  endif()
endforeach()
```

Comment (lpawelcz) [reply gecko/CMakeLists.txt]: Right, that will allow us to handle out-of-tree development. Thank you for this, I added the change to the PR.

**Review by fkokosinski — APPROVED** [review]

**Review by sateeshkotapati — APPROVED** [review]
Looks good to me

**Pull request merged**